diesel/src/insertable.rs

```diff
@@ -36,33 +38,41 @@ pub enum ColumnInsertValue<Col, Expr> where
     Default(Col),
 }
 
+type ValuesFn<Item, T, DB> = fn(Item) -> <Item as Insertable<T, DB>>::Values;
+
 impl<Iter, T, DB> Insertable<T, DB> for Iter where
     T: Table,
     DB: Backend + SupportsDefaultKeyword,
     Iter: IntoIterator,
     Iter::Item: Insertable<T, DB>,
+    Iter::IntoIter: Clone,
 {
-    type Values = BatchInsertValues<<Iter::Item as Insertable<T, DB>>::Values>;
+    type Values = BatchInsertValues<iter::Map<
+        Iter::IntoIter,
+        ValuesFn<Iter::Item, T, DB>,
+    >>;
 
     fn values(self) -> Self::Values {
-        let values = self.into_iter().map(Insertable::values).collect();
+        let values = self.into_iter()
+            .map(Insertable::values as ValuesFn<Iter::Item, T, DB>);
         BatchInsertValues(values)
     }
 }
 
 #[derive(Debug, Clone)]
-pub struct BatchInsertValues<T>(Vec<T>);
+pub struct BatchInsertValues<T>(T);
 
 impl<T, DB> InsertValues<DB> for BatchInsertValues<T> where
-    T: InsertValues<DB>,
+    T: Iterator + Clone,
+    T::Item: InsertValues<DB>,
     DB: Backend,
 {
     fn column_names(&self, out: &mut DB::QueryBuilder) -> BuildQueryResult {
-        self.0[0].column_names(out)
+        self.0.clone().next().unwrap().column_names(out)
```

Comment:
Clippy will not like this unwrap and I will probably make it a `….ok_or(WeirdStuffHappend).and_then(|val| val.column_names(out))` somewhere along the line ;)